File: core/src/main/java/com/bealetech/metrics/reporting/StatsdReporter.java
package com.bealetech.metrics.reporting;

import com.codahale.metrics.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Locale;
import java.util.Map;
import java.util.SortedMap;
import java.util.concurrent.TimeUnit;

/**
 * A reporter which publishes metric values to a Statds server.
 *
 * @see <a href="https://github.com/etsy/statsd">Statsd</a>
 */
public class StatsdReporter extends ScheduledReporter {

    /**
     * Returns a new {@link Builder} for {@link StatsdReporter}.
     *
     * @param registry the registry to report
     * @return a {@link Builder} instance for a {@link StatsdReporter}
     */
    public static Builder forRegistry(MetricRegistry registry) {
        return new Builder(registry);
    }

    /**
     * A builder for {@link StatsdReporter} instances. Defaults to not using a prefix, using the
     * default clock, converting rates to events/second, converting durations to milliseconds, and
     * not filtering metrics.
     */
    public static class Builder {
        private final MetricRegistry registry;
        private String prefix;
        private TimeUnit rateUnit;
        private TimeUnit durationUnit;
        private MetricFilter filter;

        private Builder(MetricRegistry registry) {
            this.registry = registry;
            this.prefix = null;
            this.rateUnit = TimeUnit.SECONDS;
            this.durationUnit = TimeUnit.MILLISECONDS;
            this.filter = MetricFilter.ALL;
        }

        /**
         * Prefix all metric names with the given string.
         *
         * @param prefix the prefix for all metric names
         * @return {@code this}
         */
        public Builder prefixedWith(String prefix) {
            this.prefix = prefix;
            return this;
        }

        /**
         * Convert rates to the given time unit.
         *
         * @param rateUnit a unit of time
         * @return {@code this}
         */
        public Builder convertRatesTo(TimeUnit rateUnit) {
            this.rateUnit = rateUnit;
            return this;
        }

        /**
         * Convert durations to the given time unit.
         *
         * @param durationUnit a unit of time
         * @return {@code this}
         */
        public Builder convertDurationsTo(TimeUnit durationUnit) {
            this.durationUnit = durationUnit;
            return this;
        }

        /**
         * Only report metrics which match the given filter.
         *
         * @param filter a {@link MetricFilter}
         * @return {@code this}
         */
        public Builder filter(MetricFilter filter) {
            this.filter = filter;
            return this;
        }

        /**
         * Builds a {@link StatsdReporter} with the given properties, sending metrics using the
         * given {@link Statsd} client.
         *
         * @param statsd a {@link Statsd} client
         * @return a {@link StatsdReporter}
         */
        public StatsdReporter build(Statsd statsd) {
            return new StatsdReporter(registry,
                    statsd,
                    prefix,
                    filter,
                    rateUnit,
                    durationUnit);
        }
    }

    private static final Logger LOGGER = LoggerFactory.getLogger(StatsdReporter.class);

    private final Statsd statsd;
    private final String prefix;

    public StatsdReporter(MetricRegistry registry,
                            Statsd statsd,
                            String prefix,
                            MetricFilter filter,
                            TimeUnit rateUnit,
                            TimeUnit durationUnit) {
        super(registry, "statsd-reporter", filter, rateUnit, durationUnit);

        this.statsd = statsd;
        this.prefix = prefix;
    }

    @Override
    public void report(SortedMap<String, Gauge> gauges,
                       SortedMap<String, Counter> counters,
                       SortedMap<String, Histogram> histograms,
                       SortedMap<String, Meter> meters,
                       SortedMap<String, Timer> timers) {

        try {
            statsd.connect();

            for (Map.Entry<String, Gauge> entry : gauges.entrySet()) {
                reportGauge(entry.getKey(), entry.getValue());
            }

            for (Map.Entry<String, Counter> entry : counters.entrySet()) {
                reportCounter(entry.getKey(), entry.getValue());
            }

            for (Map.Entry<String, Histogram> entry : histograms.entrySet()) {
                reportHistogram(entry.getKey(), entry.getValue());
            }

            for (Map.Entry<String, Meter> entry : meters.entrySet()) {
                reportMetered(entry.getKey(), entry.getValue());
            }

            for (Map.Entry<String, Timer> entry : timers.entrySet()) {
                reportTimer(entry.getKey(), entry.getValue());
            }

        } catch(IOException e) {
            LOGGER.warn("Unable to report to StatsD", statsd, e);
        } finally {
            try {
                statsd.close();
            } catch (IOException e) {
                LOGGER.debug("Error disconnecting from StatsD server", statsd, e);
            }
        }
    }

    private void reportTimer(String name, Timer timer) throws IOException {
        final Snapshot snapshot = timer.getSnapshot();

        statsd.send(prefix(name, "max"),
                format(convertDuration(snapshot.getMax())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "mean"),
                format(convertDuration(snapshot.getMean())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "min"),
                format(convertDuration(snapshot.getMin())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "stddev"),
                format(convertDuration(snapshot.getStdDev())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p50"),
                format(convertDuration(snapshot.getMedian())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p75"),
                format(convertDuration(snapshot.get75thPercentile())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p95"),
                format(convertDuration(snapshot.get95thPercentile())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p98"),
                format(convertDuration(snapshot.get98thPercentile())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p99"),
                format(convertDuration(snapshot.get99thPercentile())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p999"),
                format(convertDuration(snapshot.get999thPercentile())),
                Statsd.StatType.TIMER);

        reportMetered(name, timer);
    }

    private void reportMetered(String name, Metered meter) throws IOException {
        statsd.send(prefix(name, "count"), format(meter.getCount()), Statsd.StatType.GAUGE);
        statsd.send(prefix(name, "m1_rate"),
                format(convertRate(meter.getOneMinuteRate())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "m5_rate"),
                format(convertRate(meter.getFiveMinuteRate())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "m15_rate"),
                format(convertRate(meter.getFifteenMinuteRate())),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "mean_rate"),
                format(convertRate(meter.getMeanRate())),
                Statsd.StatType.TIMER);
    }

    private void reportHistogram(String name, Histogram histogram) throws IOException {
        final Snapshot snapshot = histogram.getSnapshot();
        statsd.send(prefix(name, "count"),
                format(histogram.getCount()),
                Statsd.StatType.GAUGE);
        statsd.send(prefix(name, "max"),
                format(snapshot.getMax()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "mean"),
                format(snapshot.getMean()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "min"),
                format(snapshot.getMin()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "stddev"),
                format(snapshot.getStdDev()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p50"),
                format(snapshot.getMedian()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p75"),
                format(snapshot.get75thPercentile()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p95"),
                format(snapshot.get95thPercentile()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p98"),
                format(snapshot.get98thPercentile()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p99"),
                format(snapshot.get99thPercentile()),
                Statsd.StatType.TIMER);
        statsd.send(prefix(name, "p999"),
                format(snapshot.get999thPercentile()),
                Statsd.StatType.TIMER);
    }

    private void reportCounter(String name, Counter counter) throws IOException {
        statsd.send(prefix(name, "count"),
                format(counter.getCount()),
                Statsd.StatType.COUNTER);
    }

    private void reportGauge(String name, Gauge gauge) throws IOException {
        final String value = format(gauge.getValue());
        if (value != null) {
            statsd.send(prefix(name), value,
                    Statsd.StatType.GAUGE);
        }
    }

    private String format(Object o) {
        if (o instanceof Float) {
            return format(((Float) o).doubleValue());
        } else if (o instanceof Double) {
            return format(((Double) o).doubleValue());
        } else if (o instanceof Byte) {
            return format(((Byte) o).longValue());
        } else if (o instanceof Short) {
            return format(((Short) o).longValue());
        } else if (o instanceof Integer) {
            return format(((Integer) o).longValue());
        } else if (o instanceof Long) {
            return format(((Long) o).longValue());
        }
        return null;
    }

    private String prefix(String... components) {
        return MetricRegistry.name(prefix, components);
    }

    private String format(long n) {
        return Long.toString(n);
    }

    private String format(double v) {
        return String.format(Locale.US, "%2.2f", v);
    }
}

File: core/src/main/scala/org/apache/spark/metrics/sink/StatsdReporter.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.metrics.sink

import java.io.IOException
import java.net.{DatagramPacket, DatagramSocket, InetSocketAddress}
import java.nio.charset.StandardCharsets.UTF_8
import java.util.SortedMap
import java.util.concurrent.TimeUnit

import scala.jdk.CollectionConverters._
import scala.util.{Failure, Success, Try}

import com.codahale.metrics._
import org.apache.hadoop.net.NetUtils

import org.apache.spark.internal.Logging

/**
 * @see <a href="https://github.com/etsy/statsd/blob/master/docs/metric_types.md">
 *        StatsD metric types</a>
 */
private[spark] object StatsdMetricType {
  val COUNTER = "c"
  val GAUGE = "g"
  val TIMER = "ms"
  val Set = "s"
}

private[spark] class StatsdReporter(
    registry: MetricRegistry,
    host: String = "127.0.0.1",
    port: Int = 8125,
    prefix: String = "",
    filter: MetricFilter = MetricFilter.ALL,
    rateUnit: TimeUnit = TimeUnit.SECONDS,
    durationUnit: TimeUnit = TimeUnit.MILLISECONDS)
  extends ScheduledReporter(registry, "statsd-reporter", filter, rateUnit, durationUnit)
  with Logging {

  import StatsdMetricType._

  private val address = new InetSocketAddress(host, port)
  private val whitespace = "[\\s]+".r

  override def report(
      gauges: SortedMap[String, Gauge[_]],
      counters: SortedMap[String, Counter],
      histograms: SortedMap[String, Histogram],
      meters: SortedMap[String, Meter],
      timers: SortedMap[String, Timer]): Unit =
    Try(new DatagramSocket) match {
      case Failure(ioe: IOException) => logWarning("StatsD datagram socket construction failed",
        NetUtils.wrapException(host, port, NetUtils.getHostname(), 0, ioe))
      case Failure(e) => logWarning("StatsD datagram socket construction failed", e)
      case Success(s) =>
        implicit val socket = s
        val localAddress = Try(socket.getLocalAddress).map(_.getHostAddress).getOrElse(null)
        val localPort = socket.getLocalPort
        Try {
          gauges.entrySet.asScala.foreach(e => reportGauge(e.getKey, e.getValue))
          counters.entrySet.asScala.foreach(e => reportCounter(e.getKey, e.getValue))
          histograms.entrySet.asScala.foreach(e => reportHistogram(e.getKey, e.getValue))
          meters.entrySet.asScala.foreach(e => reportMetered(e.getKey, e.getValue))
          timers.entrySet.asScala.foreach(e => reportTimer(e.getKey, e.getValue))
        } recover {
          case ioe: IOException =>
            logDebug(s"Unable to send packets to StatsD", NetUtils.wrapException(
              address.getHostString, address.getPort, localAddress, localPort, ioe))
          case e: Throwable => logDebug(s"Unable to send packets to StatsD at '$host:$port'", e)
        }
        Try(socket.close()) recover {
          case ioe: IOException =>
            logDebug("Error when close socket to StatsD", NetUtils.wrapException(
              address.getHostString, address.getPort, localAddress, localPort, ioe))
          case e: Throwable => logDebug("Error when close socket to StatsD", e)
        }
    }

  private def reportGauge(name: String, gauge: Gauge[_])(implicit socket: DatagramSocket): Unit =
    formatAny(gauge.getValue).foreach(v => send(fullName(name), v, GAUGE))

  private def reportCounter(name: String, counter: Counter)(implicit socket: DatagramSocket): Unit =
    send(fullName(name), format(counter.getCount), COUNTER)

  private def reportHistogram(name: String, histogram: Histogram)
      (implicit socket: DatagramSocket): Unit = {
    val snapshot = histogram.getSnapshot
    send(fullName(name, "count"), format(histogram.getCount), GAUGE)
    send(fullName(name, "max"), format(snapshot.getMax), TIMER)
    send(fullName(name, "mean"), format(snapshot.getMean), TIMER)
    send(fullName(name, "min"), format(snapshot.getMin), TIMER)
    send(fullName(name, "stddev"), format(snapshot.getStdDev), TIMER)
    send(fullName(name, "p50"), format(snapshot.getMedian), TIMER)
    send(fullName(name, "p75"), format(snapshot.get75thPercentile), TIMER)
    send(fullName(name, "p95"), format(snapshot.get95thPercentile), TIMER)
    send(fullName(name, "p98"), format(snapshot.get98thPercentile), TIMER)
    send(fullName(name, "p99"), format(snapshot.get99thPercentile), TIMER)
    send(fullName(name, "p999"), format(snapshot.get999thPercentile), TIMER)
  }

  private def reportMetered(name: String, meter: Metered)(implicit socket: DatagramSocket): Unit = {
    send(fullName(name, "count"), format(meter.getCount), GAUGE)
    send(fullName(name, "m1_rate"), format(convertRate(meter.getOneMinuteRate)), TIMER)
    send(fullName(name, "m5_rate"), format(convertRate(meter.getFiveMinuteRate)), TIMER)
    send(fullName(name, "m15_rate"), format(convertRate(meter.getFifteenMinuteRate)), TIMER)
    send(fullName(name, "mean_rate"), format(convertRate(meter.getMeanRate)), TIMER)
  }

  private def reportTimer(name: String, timer: Timer)(implicit socket: DatagramSocket): Unit = {
    val snapshot = timer.getSnapshot
    send(fullName(name, "max"), format(convertDuration(snapshot.getMax.toDouble)), TIMER)
    send(fullName(name, "mean"), format(convertDuration(snapshot.getMean)), TIMER)
    send(fullName(name, "min"), format(convertDuration(snapshot.getMin.toDouble)), TIMER)
    send(fullName(name, "stddev"), format(convertDuration(snapshot.getStdDev)), TIMER)
    send(fullName(name, "p50"), format(convertDuration(snapshot.getMedian)), TIMER)
    send(fullName(name, "p75"), format(convertDuration(snapshot.get75thPercentile)), TIMER)
    send(fullName(name, "p95"), format(convertDuration(snapshot.get95thPercentile)), TIMER)
    send(fullName(name, "p98"), format(convertDuration(snapshot.get98thPercentile)), TIMER)
    send(fullName(name, "p99"), format(convertDuration(snapshot.get99thPercentile)), TIMER)
    send(fullName(name, "p999"), format(convertDuration(snapshot.get999thPercentile)), TIMER)

    reportMetered(name, timer)
  }

  private def send(name: String, value: String, metricType: String)
      (implicit socket: DatagramSocket): Unit = {
    val bytes = sanitize(s"$name:$value|$metricType").getBytes(UTF_8)
    val packet = new DatagramPacket(bytes, bytes.length, address)
    socket.send(packet)
  }

  private def fullName(names: String*): String = MetricRegistry.name(prefix, names : _*)

  private def sanitize(s: String): String = whitespace.replaceAllIn(s, "-")

  private def format(v: Any): String = formatAny(v).getOrElse("")

  private def formatAny(v: Any): Option[String] =
    v match {
      case f: Float => Some("%2.2f".format(f))
      case d: Double => Some("%2.2f".format(d))
      case b: BigDecimal => Some("%2.2f".format(b))
      case n: Number => Some(v.toString)
      case _ => None
    }
}

